### ISSUE TITLE
Support dependency without outputs

### ISSUE THREAD

**Issue by damon-atkins** [gruntwork-io/terragrunt#1499]
terragrunt_v0.26.7 calls dependency with  terraform output -json which response with
```
{}
```

Terragrunts messages are
```
[terragrunt] [dir] 2021/01/20 18:14:13 Running command: terraform output -json
[terragrunt] 2021/01/20 18:14:37 xxxxxxx/terragrunt.hcl is a dependency of  yyyyy/terragrunt.hcl but detected no outputs. Either the target module has not been applied yet, or the module has no outputs. If this is expected, set the skip_outputs flag to true on the dependency block.
[terragrunt] 2021/01/20 18:14:37 Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```
Consider `{}` a valid response of no `output` however error as its expected to have output without waiting 30 seconds.

**Comment by brikis98** [comment]
Any chance you could share the code so we have some way to repro this?

**Comment by yorinasub17** [comment]
After looking at this again, I think the issue here is that terragrunt `dependency` always assumes there is some output provided by the module, since that is the point of the explicit dependency (to string outputs to the dependent module).

I think the feature request here is to support this semantic of `dependency` where you only want to add it for the purpose of deployment ordering with `xxx-all` commands. I updated the issue to reflect this.

For now, the workaround is to use the `dependencies` block for these modules instead, which is only used for the purpose of dependency ordering as opposed to dependency passing.

**Comment by yorinasub17** [comment]
Another workaround: if you know you are not using any outputs from `dependency` block, you can also set `skip_outputs = true`. With `skip_outputs = true`, terragrunt would still include the dependency in the graph, but will skip pulling out the outputs with `terragrunt output`, thereby bypassing the check for empty output.

**Comment by damon-atkins** [comment]
It seems when terraform outputs  `{}` only "terragrunt appears to timeout waiting for the output" instead of moving on. I am surprised its hitting the timeout limit.    Maybe its looking  for
 ```
{
 }
```
When their  is no output variables in json

**Comment by damon-atkins** [comment]
It **could** also be that terraform has changed its behavior when their are no outputs and you request json
```
$ terraform output

Warning: No outputs found

The state file either has no outputs defined, or all the defined outputs are
empty. Please define an output in your configuration with the `output` keyword
and run `terraform refresh` for it to become available. If you are using
interpolation, please verify the interpolated value is not empty. You can use
the `terraform console` command to assist.

$ terraform output -json
{}
```

**Comment by damon-atkins** [comment]
@yorinasub17  please label this as a bug, as terragrunt is stuck because terraform outputs  `{}` when terragrunt expects `{<linefeed>}`

**Comment by brikis98** [comment]
@denis256 could you look into this one?

**Comment by denis256** [comment]
Hello,
with terraform 1.0.x, stuck issue is not happening - tested with terragrunt `v0.26.7` and with the version from master `0.31.2`, now terraform outputs json variables as `{}` which is handled correctly.

**Comment by brikis98** [comment]
@denis256 What happens in your test repo if `app` declares no `output` variables whatsoever?

**Comment by denis256** [comment]
With commented `output`,  immediate error response:
```
$ terragrunt init
ERRO[0000] app/terragrunt.hcl is a dependency of terragrunt.hcl but detected no outputs. Either the target module has not been applied yet, or the module has no outputs. If this is expected, set the skip_outputs flag to true on the dependency block. 
ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

**Comment by damon-atkins** [comment]
I suspect the issue is when terraform outputs is:
 ``
{}
``
vs expect by the terragrunt code is
``
{
}
``
When the JSON contains something there are line feeds after the first `{` and before the last `}` 
``
{
   something = "hello"
}
``

**Comment by denis256** [comment]
True, but this is not happening now on the last terraform version 1.0.x), I can try to hack maybe invocation of `terraform output -json ` to add a new line to see if the behaviour will be different

**Comment by damon-atkins** [comment]
I suspect terraform would have been 0.13 or 0.14. I was hoping you could read the terragrunt code to check it handles just getting  `{}` from output. Sorry not a `go` programmer.

**Comment by damon-atkins** [comment]
Terragrunt handles it with a timeout, rather than knowing its output is finished.

**Comment by damon-atkins** [comment]
Thanks for looking into this.....

**Comment by brikis98** [comment]
> With commented `output`, immediate error response:
> 
> ```
> $ terragrunt init
> ERRO[0000] app/terragrunt.hcl is a dependency of terragrunt.hcl but detected no outputs. Either the target module has not been applied yet, or the module has no outputs. If this is expected, set the skip_outputs flag to true on the dependency block. 
> ERRO[0000] Unable to determine underlying exit code, so Terragrunt will exit with error code 1
> ```
> 
> ![Screenshot at 2021-09-03 12-49-38](https://user-images.githubusercontent.com/10694338/131986460-98f78109-2fe7-4eda-a5f7-c14b39c0c78f.png)

Right, so I think the request in this issue is that Terragrunt does not error out when outputs are empty, but allows it.

**Comment by damon-atkins** [comment]
Had a look though the source code and can not see any reason for a 20 second delay.

```
terraform output  ; echo $?
╷
│ Warning: No outputs found
│
│ The state file either has no outputs defined, or all the defined outputs are
│ empty. Please define an output in your configuration with the `output`
│ keyword and run `terraform refresh` for it to become available. If you are
│ using interpolation, please verify the interpolated value is not empty. You
│ can use the `terraform console` command to assist.
╵
0
```

```
$ terraform output -json ; echo $?
{}
0
```
Note that to me `Unable to determine underlying exit code`  (main.go) reads to me that terragrunt did not get an exit code from terraform.  Is that the intention of this message?

**Comment by brikis98** [comment]
This should've been fixed by https://github.com/gruntwork-io/terragrunt/pull/1811.

**Comment by damon-atkins** [comment]
Thanks 😃

**Comment by yorinasub17** [comment]
I think we are going to have to revert #1811 because it had the unfortunate side effect of producing an unhelpful error message when the output is empty. Specifically, in the following example:

```
dependency "dep" {
  config_path = "../dep"
}

inputs = {
  input = dependency.dep.outputs.id
}
```

The error message you get when you run `plan` without the dependency being applied, is:

```
ERRO[0002] Module ./main has finished with an error: ./main/terragrunt.hcl:6,25-33: Unsupported attribute; This object does not have an attribute named "outputs".  prefix=[./main]
```

which is very confusing because it implies the `outputs` syntax is wrong, not that there is no outputs from the dependency which is what the original error message indicated.

I don't actually see a use case to allow empty outputs with `dependency` without a config that indicates no output should be fetched (which is the `skip_outputs = true` flag), as otherwise, the `terragrunt.hcl` would reference `dependency.dep.outputs` and empty outputs will fail.

---

For the issue at hand, since there is a concern that reverting that PR might cause the timeout to happen again:

Looking at the code, the only place where this 30 second "timeout" might come into play is [this line](https://github.com/gruntwork-io/terragrunt/blob/master/config/dependency.go#L289), where it is waiting for all the dependency fetches to finish. Is it possible that the 30 second "timeout" is actually terragrunt waiting for other dependency fetching to happen in parallel, one of which might be using remote state and there is a network delay? This might be highlighted more with verbose logging using `--terragrunt-log-level debug`.

Another possibility is that the "timeout" is actually from terragrunt running the `output` command. Under certain circumstances, note that `terragrunt output` could take a long time because it runs `terragrunt init` on the dependent module to initialize the remote state (see the [docs on dependency](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#dependency), specifically section "Can I speed up dependency fetching?"). You can disable this and use the cached `init` if you set `disable_dependency_optimization = true` on the `remote_state` block.

The reason I don't think the issue is that terragrunt is waiting for a line feed, is because the terraform output logic is not parsing the output line by line. In fact, the only wait that happens regarding the outputs is the wait for the terraform subprocess to finish (it uses this function under the hood [RunTerraformCommandWithOutput](https://github.com/gruntwork-io/terragrunt/blob/eaa3028c80e7b57461112c08f78998a6b5db0596/shell/run_shell_cmd.go#L42), where the only wait is on `cmd.Wait`), so it doesn't matter if terraform returns `{}` or `{<linefeed>}`.

Either way, the fix in #1811 wouldn't make a difference to the timeout logic because the code change is not directly related to the wait logic in terragrunt. So I think reverting it shouldn't have an effect on the wait times.

**Comment by denis256** [comment]
Hm, ok, I agree to revert this change if it is generating more confusions
